src/compiler/parser.ts

```diff
@@ -1762,10 +1763,17 @@ namespace Parser {
         return hasJSDoc ? addJSDocComment(node) : node;
     }
 
+    const seeLink = /@(?:see|link)/;
+    function shouldParseJSDoc<T extends HasJSDoc>(node: T, comment: ts.CommentRange) {
+        if (!(globalThis as any).isTSC) return true;
+        if (node.flags & NodeFlags.JavaScriptFile) return true;
+        if (seeLink.test(sourceText.slice(comment.pos, comment.end))) return true;
+    }
+
     let hasDeprecatedTag = false;
     function addJSDocComment<T extends HasJSDoc>(node: T): T {
         Debug.assert(!node.jsDoc); // Should only be called once per node
-        const jsDoc = mapDefined(getJSDocCommentRanges(node, sourceText), comment => JSDocParser.parseJSDocComment(node, comment.pos, comment.end - comment.pos));
+        const jsDoc = mapDefined(getJSDocCommentRanges(node, sourceText), comment => shouldParseJSDoc(node, comment) && JSDocParser.parseJSDocComment(node, comment.pos, comment.end - comment.pos));

```

Comment:
Did you consider having the scanner check for `@link` and `@see` in the code in `scan()` that already checks for JSDoc and sets the `TokenFlags.PrecedingJSDocComment` flag. This could either be a new flag, or you could simply say that in `.ts` files you only set the `TokenFlags.PrecedingJSDocComment` when there are `@link` and `@see` tags. Then `withJSDoc` would completely skip over this additional scanning which should save you even more time.